src/Storages/WindowView/StorageWindowView.cpp

```diff
@@ -0,0 +1,1401 @@
+{
+    IntervalKind window_interval_kind;
+    if (is_tumble)
+        window_interval_kind = window_kind;
+    else
+        window_interval_kind = hop_kind;
+
+    switch (window_interval_kind)
+    {
+#define CASE_WINDOW_KIND(KIND) \
+    case IntervalKind::KIND: \
+    { \
+        if (is_tumble) \
+            return ToStartOfTransform<IntervalKind::KIND>::execute(time_sec, window_num_units, *time_zone); \
+        else \
+        {\
+            UInt32 w_start = ToStartOfTransform<IntervalKind::KIND>::execute(time_sec, hop_num_units, *time_zone); \
+            UInt32 w_end = AddTime<IntervalKind::KIND>::execute(w_start, hop_num_units, *time_zone);\
+            return AddTime<IntervalKind::KIND>::execute(w_end, -1 * window_num_units, *time_zone);\
+        }\
+    }
+        CASE_WINDOW_KIND(Second)
+        CASE_WINDOW_KIND(Minute)
+        CASE_WINDOW_KIND(Hour)
+        CASE_WINDOW_KIND(Day)
+        CASE_WINDOW_KIND(Week)
+        CASE_WINDOW_KIND(Month)
+        CASE_WINDOW_KIND(Quarter)
+        CASE_WINDOW_KIND(Year)
+#undef CASE_WINDOW_KIND
+    }
+    __builtin_unreachable();
+}
+
+UInt32 StorageWindowView::getWindowUpperBound(UInt32 time_sec)
+{
+    IntervalKind window_interval_kind;
+    if (is_tumble)
+        window_interval_kind = window_kind;
+    else
+        window_interval_kind = hop_kind;
+
+    switch (window_interval_kind)
+    {
+#define CASE_WINDOW_KIND(KIND) \
+    case IntervalKind::KIND: \
+    { \
+        if (is_tumble) \
+        {\
+            UInt32 w_start = ToStartOfTransform<IntervalKind::KIND>::execute(time_sec, window_num_units, *time_zone); \
+            return AddTime<IntervalKind::KIND>::execute(w_start, window_num_units, *time_zone); \
+        }\
+        else \
+        {\
+            UInt32 w_start = ToStartOfTransform<IntervalKind::KIND>::execute(time_sec, hop_num_units, *time_zone); \
+            return AddTime<IntervalKind::KIND>::execute(w_start, hop_num_units, *time_zone); \
+        }\
+    }
+        CASE_WINDOW_KIND(Second)
+        CASE_WINDOW_KIND(Minute)
+        CASE_WINDOW_KIND(Hour)
+        CASE_WINDOW_KIND(Day)
+        CASE_WINDOW_KIND(Week)
+        CASE_WINDOW_KIND(Month)
+        CASE_WINDOW_KIND(Quarter)
+        CASE_WINDOW_KIND(Year)
+#undef CASE_WINDOW_KIND
+    }
+    __builtin_unreachable();
+}
+
+void StorageWindowView::addFireSignal(std::set<UInt32> & signals)
+{
+    std::lock_guard lock(fire_signal_mutex);
+    for (const auto & signal : signals)
+        fire_signal.push_back(signal);
+    fire_signal_condition.notify_all();
+}
+
+void StorageWindowView::updateMaxTimestamp(UInt32 timestamp)
+{
+    std::lock_guard lock(fire_signal_mutex);
+    if (timestamp > max_timestamp)
+        max_timestamp = timestamp;
+}
+
+void StorageWindowView::updateMaxWatermark(UInt32 watermark)
+{
+    std::lock_guard lock(fire_signal_mutex);
+    if (max_watermark == 0)
+    {
+        max_watermark = getWindowUpperBound(watermark - 1); /// TODO: why -1?
```

Comment:
The `max_watermark` should be the next watermark to be fired. For a tumbling window with 3 seconds size as follows: 
```
|           |           |
|___|___|___|___|___|___|
0   1   2   3   4   5   6
```
If we insert a new data block with three timestamps 1, 2, 3. The `watermark` in line 757 is 3. We add -1 to set the `max_watermark` to 3 instead of 6 and avoid losing the first window's data (1, 2)

Reply:
Got it, thanks